### ISSUE TITLE
Rewriting `.ts` module specifiers to `.js` without a plugin

### ISSUE THREAD

**Issue by bakkot** [evanw/esbuild#2435]
The TypeScript team seems set on making a decision I think is going to lead to a lot of pain for beginners: they are categorically against rewrite module specifiers. This means that beginners who are using `tsc` as a compiler are going to need to deal with the [super hacky](https://twitter.com/mattgodbolt/status/1543325727663218690) thing where you write `import "foo.js"` to refer to a file named `foo.ts`. I think that's going to cause a huge amount of confusion, over and above the existing pain the esm transition entails. It already is, really. (I was somewhat distressed, although not altogether surprised, to learn from the above-linked tweet that the hard part of building [godbolt.org](https://godbolt.org) is apparently, somehow, _JavaScript_.)

But it looks like [TypeScript might allow you to write `.ts` in module specifiers](https://github.com/microsoft/TypeScript/issues/37582#issuecomment-1184950658) when using `noEmit`, relying on other tools to do that transformation for you. I'm hoping `esbuild` can be such a tool _without needing a plugin_.

I know `esbuild` is already [capable of doing rewrites of module specifiers using a plugin](https://github.com/evanw/esbuild/issues/622#issuecomment-769462611), but the plugin API is a lot to ask a beginner to deal with. (Plus it requires enabling bundling even though you are actually not bundling at all, which is confusing in its own right.)

So: I'd like to ask for `esbuild` to support rewriting extensions in module specifiers without needing a plugin. That way we as a community would have a reasonable answer for a beginner setting out to do the should-be-straightforward task of writing a simple node.js application with TypeScript: 1) write module specifiers the way you expect to write them, i.e. with a `.ts` extension, 2) use `tsc --noEmit --strict` for type checking, 3) use `esbuild --rewrite-ts` (or whatever) to actually create `.js` files which you can run.

Again, I know it's possible to accomplish this with a plugin, but I'm concerned specifically about the experience for people new to JavaScript, trying to do something simple; I think a plugin is a bad match for that specific case. And again I realize this is a divergence from TypeScript, but that's precisely the point; TypeScript has decided that they are willing to sacrifice the experience of beginners, and I'm hoping esbuild might decide otherwise, so that as a community we can have at least _one_ reasonable option for people just getting started.

**Comment by hyrious** [comment]
First of all, esbuild is not a transpiler like swc or babel at the first place, it is a bundler and it can already make sure multiple chunks could refer to each other using the right extension. [REPL Link](https://hyrious.me/esbuild-repl/?version=0.14.51&mode=build&modules=%5B%5B%22main.js%22%2C%22export+*+from+%5C%22.%2Flib%5C%22%22%2C1%5D%2C%5B%22lib.js%22%2C%22export+let+a+%3D+1%22%2C1%5D%5D&buildOptions=%7B%22bundle%22%3Atrue%2C%22format%22%3A%22esm%22%2C%22splitting%22%3Atrue%2C%22outExtension%22%3A%7B%22.js%22%3A%22.mjs%22%7D%7D)

Second, I guess you're asking esbuild to automatically transform _external_ paths to replace `.ts` with `.js`? Like, if I have this input:

```js
import "./lib.ts"
```

With `esbuild --bundle --external:./b.ts --format=esm`, you're expecting:

```js
import "./lib.js"
```

Correct me if I'm wrong.

If that's true, then I don't think this would be any useful as normally people won't externalize one local file in this way. And again esbuild is not designed in the first place to transpile files like babel/swc/[rollup](https://rollupjs.org/guide/en/#outputpreservemodules). I can understand that some node loaders using esbuild would expect improving the transform api to do this work.

**Comment by bakkot** [comment]
> esbuild is not a transpiler like swc or babel at the first place

esbuild is in fact a transpiler? I'm not sure what claim you're making here. esbuild without `--bundle` will strip typescript types, and adding `--format=cjs` will translate esm to cjs.

>  I guess you're asking esbuild to automatically transform _external_ paths to replace .ts with .js?

I'm asking about use _without_ `--bundle` - i.e., the default behavior, which mostly just strips TypeScript types. So, yes, in a sense I'm asking about automatically transforming external paths, but where all paths are external.

> If that's true, then I don't think this would be any useful as normally people won't externalize one local file in this way

If you have to name every individual file in `--external`, I agree people won't do that, but that's only necessary when using `--bundle`. What I'm asking about is the case where `esbuild` is just being used to strip types (and possibly convert `esm` to commonjs). This is already something people do, since `esbuild` is _much_ faster than `tsc` for this application.

**Comment by Conaclos** [comment]
TypeScript 5.0 beta supports imports of files with `.ts` extension in `bundler` mode.

I could refine the proposal of [bakkot](https://github.com/bakkot) as follows:

Change the extension of `.ts` to `.js` for imports that are part of entry points.
External imports should not be rewritten because a custom (TypeScript) loader may be used.

  e.g.

  ```ts
  $ more src/internal.ts
  export * from "some/external.ts"

  $ more src/index.ts
  import { X } from "./internal.ts"

  $ esbuild src/*.ts --outdir=dist
  $ more dist/internal.js
  export * from "some/external.ts"

  $ more dist/index.js
  import { X } from "./internal.js"
  ```

**Comment by Conaclos** [comment]
This may be an issue worth reconsidering, as [TypeScript 5.7 will ship with a new option](https://github.com/microsoft/TypeScript/pull/59767) to rewrite relative import with `ts` extension to relative import with `js` extension.

**Comment by gulshan** [comment]
Node.js now supports running .ts files directly, but the to import typescript files, the extension has to be `.ts`. esbuild should now have an option to rewrite `.ts` to `.js`.

**Comment by evanw** [comment]
Sorry, I don't understand. If Node lets you import `.ts` files but the import must be `.ts`, then wouldn't changing `.ts` to `.js` break import of `.ts` files?

**Comment by bakkot** [comment]
The state of things has changed substantially since I wrote the original issue. Now node supports running TypeScript files directly and requires you to use `".ts"` in your imports. So the case I was originally worried about is now well supported: a beginner writing an app can just use TypeScript and write imports in the obvious way and it will all work.

However, node only supports this for things outside of `node_modules`. So if you're writing an app which is going to be published on npm, then you'll still need to strip types (and in the process rename the files from `.ts` to `.js`) and you'll also need to rewrite `".ts"` in imports to `".js"` to match the renamed files.

TypeScript could always do the type stripping part and can now also do the specifier rewriting with their new(-ish) `rewriteRelativeImportExtensions` flag, as detailed in [this PR](https://github.com/microsoft/TypeScript/pull/59767) and [the release notes](https://devblogs.microsoft.com/typescript/announcing-typescript-5-7/#path-rewriting-for-relative-paths). It would be nice if esbuild could do the same.

**Comment by sant123** [comment]
Meanwhile this gets solved, I have created [esbuild-plugin-rewrite-relative-import-extensions](https://jsr.io/@onyx/esbuild-plugin-rewrite-relative-import-extensions).

**Comment by indutny-signal** [comment]
Our use case is similar, yet different. We use `esbuild` in transpiler mode (without `--bundle` CLI argument). Our dilemma is that we can either:

- Specify all imports with `.js` extension, but then `esbuild` will prefer the last generated `.js` file even though `.ts` file is right next to it, and thus we opt out of ESM tree shaking
- Specify all imports with `.ts`/`.tsx` extensions, but then transpiled files cannot be run with `node` because our `.ts` files have enums and namespaces that are not supported by Node.

I think it'd be great if we had a way to either override the resolution order to make it prefer the `.ts` files even in presence of `.js` files as specified in the import path, or to add an option to replace `.ts` extension with `.js`. Is either of these a possibility?

**Comment by evanw** [comment]
> add an option to replace `.ts` extension with `.js`

I think the most straightforward way to do this would be for esbuild to detect `rewriteRelativeImportExtensions` in `tsconfig.json` and then try to do [what TypeScript does](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-5-7.html#path-rewriting-for-relative-paths). Specifically:

> When an import path is _relative_ (starts with `./` or `../`), ends in a TypeScript extension (`.ts`, `.tsx`, `.mts`, `.cts`), and is a non-declaration file, the compiler will rewrite the path to the corresponding JavaScript extension (`.js`, `.jsx`, `.mjs`, `.cjs`).

Perhaps that should only work when not bundling, or when the path is external. Or perhaps it should happen before path resolution even when bundling. Not sure yet. In any case, would something like that work for you?

**Comment by indutny-signal** [comment]
> Perhaps that should only work when not bundling, or when the path is external. Or perhaps it should happen before path resolution even when bundling. Not sure yet. In any case, would something like that work for you?

Absolutely! Thank you so much!
